### PR TITLE
doc: update life of a search query doc to be current

### PR DESCRIPTION
**Pull request by keegancsmith**
This document is quite out of date. I spent an hour making sure all
links where correct and the description was correct. Please read over it
in full and try out the links. There is likely more detail we can add
here, especially from Rijnard who has been overhauling a lot of the
frontend related logic.

### CODE REVIEW

**Comment by stefanhengl** [comment]
Not related to the content: this doc-page is really hard to find unless you know what you are looking for. In the end I only managed to find it by searching for "life".

File: doc/dev/background-information/architecture/life-of-a-search-query.md [hunk 1]
```diff
@@ -14,11 +14,11 @@ In all cases, clients use the [search query](https://sourcegraph.com/search?q=re
 
 ## Frontend
 
-The frontend implements the GraphQL search resolver [here](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+%28r+*schemaResolver%29+Search%28).
+The frontend uses a [Streaming API](../../../api/stream_api/index.md). The frontend implements it [here](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+%28h+*streamHandler%29+ServeHTTP). The frontend implements the GraphQL search resolver [here](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+%28r+*schemaResolver%29+Search%28). The Streaming API is used by the browser. Historically we served results via GraphQL and there are still many clients who use this API. Internally Sourcegraph search is streaming based.
```
Comment (stefanhengl) [inline doc/dev/background-information/architecture/life-of-a-search-query.md:17]:
```suggestion
The frontend uses a [Streaming API](../../../api/stream_api/index.md). The frontend implements it [here](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+%28h+*streamHandler%29+ServeHTTP). The Streaming API is used by the browser. Historically we served results via [GraphQL](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+%28r+*schemaResolver%29+Search%28) and there are still many clients who use this API. Internally Sourcegraph search is streaming based.
```

File: doc/dev/background-information/architecture/life-of-a-search-query.md [hunk 2]
```diff
@@ -14,11 +14,11 @@ In all cases, clients use the [search query](https://sourcegraph.com/search?q=re
...
+Most Sourcegraph queries can be directly translated into Zoekt queries without consulting our database of repositories. However, not all repositories or revisions are indexed. So we need to work out what isn't in zoekt so we can do unindexed queries against Searcher. So the frontend [determines which repository@revision combinations are indexed by zoekt](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+zoektIndexedRepos%28+file:indexed_search.go&patternType=literal) by [consulting an in-memory cache](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+type+cachedSearcher). See [RepoSubsetTextSearch](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+type+RepoSubsetTextSearch&patternType=literal) for how searcher is queried.
```
Comment (stefanhengl) [inline doc/dev/background-information/architecture/life-of-a-search-query.md:21]:
```suggestion
Most Sourcegraph queries can be directly translated into Zoekt queries without consulting our database of repositories. However, not all repositories or revisions are indexed. So we need to work out what isn't indexed by Zoekt so we can do unindexed queries against Searcher. So the frontend [determines which repository@revision combinations are indexed by Zoekt](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+zoektIndexedRepos%28+file:indexed_search.go&patternType=literal) by [consulting an in-memory cache](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+type+cachedSearcher). See [RepoSubsetTextSearch](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+type+RepoSubsetTextSearch&patternType=literal) for how searcher is queried.
```

File: doc/dev/background-information/architecture/life-of-a-search-query.md [hunk 3]
```diff
@@ -27,8 +27,8 @@ zoekt-webserver [serves search requests](https://sourcegraph.com/search?q=contex
...
-It fetches git data by calling [another internal frontend API](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/zoekt%24+"func+tarballURL") which [redirects to the archive on gitserver](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+serveGitTar+file:internal.go&patternType=literal). If indexing multiple branches, it instead relies on [git shallow clones](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+-file:%28test%7Cspec%7Cdoc%29+GitUploadPack&patternType=literal).
+It uses relies on [git shallow clones](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+-file:%28test%7Cspec%7Cdoc%29+GitUploadPack&patternType=literal) to fetch the contents of the branches to index. It then calls out to `zoekt-git-index` which creates 1 or more shards containing the indexes used by `zoekt-webserver`.
```
Comment (stefanhengl) [inline doc/dev/background-information/architecture/life-of-a-search-query.md:30]:
```suggestion
It uses on [git shallow clones](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+-file:%28test%7Cspec%7Cdoc%29+GitUploadPack&patternType=literal) to fetch the contents of the branches to index. It then calls out to `zoekt-git-index` which creates 1 or more shards containing the indexes used by `zoekt-webserver`.
```

Comment (stefanhengl) [reply doc/dev/background-information/architecture/life-of-a-search-query.md:30]: I would have expected this link for shallow clones. Wdyt? https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/zoekt%24+-file:_test+--depth%3D1&patternType=literal

File: doc/dev/background-information/architecture/life-of-a-search-query.md [hunk 4]
```diff
@@ -14,11 +14,11 @@ In all cases, clients use the [search query](https://sourcegraph.com/search?q=re
...
+First, the frontend takes the query and [creates a plan of jobs](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+%28r+*searchResolver%29+toSearchInputs&patternType=literal) to execute concurrently. A job is a specific query against a backend. For example [converting a Sourcegraph query into a Zoekt query](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+querytozoektquery&patternType=literal) for our indexed search backend. The comments in the job creation function are worth reading for more details.
```
Comment (stefanhengl) [inline doc/dev/background-information/architecture/life-of-a-search-query.md:19]:
```suggestion
First, the frontend takes the query and [creates a plan of jobs](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+%28r+*searchResolver%29+toSearchInputs&patternType=literal) to execute concurrently. A job is a specific query against a backend. For example [here we convert a Sourcegraph query into a Zoekt query](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+querytozoektquery&patternType=literal) for our indexed search backend. The comments in the job creation function are worth reading for more details.
```

**Review by stefanhengl — APPROVED** [review]
Thanks for doing this! I remember reading this when I joined and found it really helpful. 

I believe the section "Clients" needs an update too. It just refers to the GraphQL API?

I did one pass and clicked on all links. I left a couple of suggestions every time I had to stop and read it twice.

File: doc/dev/background-information/architecture/life-of-a-search-query.md (hunk 3 above)
Comment (keegancsmith) [reply doc/dev/background-information/architecture/life-of-a-search-query.md:30]: You are right, I'll update.

**Comment by keegancsmith** [comment]
> I believe the section "Clients" needs an update too. It just refers to the GraphQL API?

@stefanhengl updated to reference both. Also feel free to make other updates to this doc

> Not related to the content: this doc-page is really hard to find unless you know what you are looking for. In the end I only managed to find it by searching for "life".

I agree. Any suggestions? Sometimes people stumble on it though, since I occasionally get a thank you in a DM on slack.

**Comment by keegancsmith** [comment]
Gonna merge in now, @rvantonder and @camdencheek feel free to further improve this doc based on the work you all have done as well as fact check my writing :) Thanks.

**Review by camdencheek — COMMENTED** [review]
Thank you for this! 

On a related note, this doc seems like a really good candidate for a search notebook 🤔

**Pull request merged**